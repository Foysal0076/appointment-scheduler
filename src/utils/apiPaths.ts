export const apiPaths = {
  register: '/api/register',
  login: '/api/login',
  logout: '/api/logout',
  appointments: '/api/appointments',
  users: '/api/users',
}
