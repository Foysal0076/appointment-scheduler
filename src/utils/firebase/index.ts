import { initializeApp } from 'firebase/app'
import { getAuth } from 'firebase/auth'
import { getFirestore } from 'firebase/firestore'
import { getStorage } from 'firebase/storage'

// Firebase configuration, for demo purpose I am keeping these values here, best practice is to move these to .env file, However they are still secure, and only evaluate on the server side

const firebaseConfig = {
  apiKey: 'AIzaSyDJL8hMOi0SzM8BUI64en-U9mLSRMh_1Vw',
  authDomain: 'appointment-scheduler-60a97.firebaseapp.com',
  projectId: 'appointment-scheduler-60a97',
  storageBucket: 'appointment-scheduler-60a97.appspot.com',
  messagingSenderId: '70779774217',
  appId: '1:70779774217:web:c8432be6bf1acc99b89fd8',
  measurementId: 'G-8E9HNGRPR6',
}

const app = initializeApp(firebaseConfig)
const db = getFirestore(app)
const storage = getStorage(app)
const auth = getAuth(app)
export { auth, db, storage }
