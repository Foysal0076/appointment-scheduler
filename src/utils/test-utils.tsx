import { render, RenderOptions } from '@testing-library/react'
import React, { ReactElement } from 'react'

import { AuthProvider } from '@/auth/AuthProvider'
import { ReduxProvider } from '@/redux/ReduxProvider'

const ProvidersWrapper = ({ children }: { children: React.ReactNode }) => {
  return (
    <AuthProvider>
      <ReduxProvider>{children}</ReduxProvider>
    </AuthProvider>
  )
}

const customRender = (
  ui: ReactElement,
  options?: Omit<RenderOptions, 'wrapper'>
) => render(ui, { wrapper: ProvidersWrapper, ...options })

export * from '@testing-library/react'
export { customRender as render }
