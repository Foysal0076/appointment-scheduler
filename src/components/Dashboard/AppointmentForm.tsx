'use client'

import 'react-datepicker/dist/react-datepicker.css'

import { useEffect, useState } from 'react'
import DatePicker from 'react-datepicker'
import Select from 'react-select'
import { toast } from 'react-toastify'

import Button from '@/components/Common/Button'
import { Input } from '@/components/Common/Input'
import OutlinedButton from '@/components/Common/OutlinedButton'
import AppointmentConfirmModal from '@/components/Dashboard/AppointmentConfirmModal'
import AudioMessage from '@/components/Dashboard/AudioMessage'
import { useAppointmentFormData } from '@/components/Dashboard/Hooks/useAppointmentFormData'
import { useFetchUsersQuery } from '@/redux/apiQueries/userQueries'
import { MEETING_DURATION_OPTIONS } from '@/utils/constants/appointment.constants'
import { filterPassedTime } from '@/utils/helpers'
import { useUserInfo } from '@/utils/hooks/useUserInfo'

const customStyles = {
  menuPortal: (base: any) => ({ ...base, zIndex: 99999 }),
}

const reactSelectTheme = (theme: any) => ({
  ...theme,
  borderRadius: '.25rem',
  colors: {
    ...theme.colors,
    primary25: '#6b7280',
  },
})

type Props = {
  userId: string | null
  onCancel: () => void
}

const AppointmentForm = ({ userId, onCancel }: Props) => {
  const [isConfirmModalOpen, setIsConfirmModalOpen] = useState(false)
  const [options, setOptions] = useState<any[]>([])

  const { data: userResponse, isLoading: isFetchingUsers } = useFetchUsersQuery(
    {}
  )

  const { user } = useUserInfo()

  const {
    formData,
    onChangeDuration,
    onChangeGuest,
    onChangeDate,
    onChangeTitle,
    onChangeDescription,
    onChangeAudioFile,
  } = useAppointmentFormData()

  const handleOpenConfirmModal = () => {
    //check form values
    if (
      formData.guest.label === '' ||
      formData.description === '' ||
      formData.date === null
    ) {
      toast.error('Please fill all the required fields')
      return
    }
    setIsConfirmModalOpen(true)
  }
  const handleCloseConfirmModal = () => setIsConfirmModalOpen(false)

  useEffect(() => {
    if (Array.isArray(userResponse)) {
      //filter current user
      const users = userResponse.filter((u) => u.id !== user?.id)
      const options = users.map((user) => ({
        label: user.fullname,
        value: user.id,
      }))
      setOptions(options)
      const selectedUser =
        options.find((option) => option.value === userId) ?? options[0]
      onChangeGuest(selectedUser)
    }
  }, [userResponse, userId])

  return (
    <div className='w-[85vw] md:max-w-lg'>
      <h2 className='h3 mb-8'>Create Meeting</h2>
      <div className='gap- flex flex-col gap-6'>
        <div className='flex flex-col'>
          <label htmlFor='meeting-title' className='mb-2 font-bold'>
            Title
          </label>
          <Input
            name='meeting-title'
            id='meeting-title'
            onChange={onChangeTitle}
            placeholder='Give your meeting a title'
          />
        </div>
        <div className='flex flex-col'>
          <label htmlFor='meeting-description' className='mb-2 font-bold'>
            Description <span className='ml-0.5 text-danger-500'>*</span>
          </label>
          <textarea
            name='meeting-description'
            id='meeting-description'
            onChange={onChangeDescription}
            placeholder='Add a description'
            rows={4}
            className='w-full resize-none rounded bg-surface-50 dark:bg-surface-200'
          />
        </div>
        <div className='flex flex-col'>
          <label htmlFor='audio-message' className='mb-2 font-bold'>
            Audio message
          </label>
          <AudioMessage
            setAudioFile={onChangeAudioFile}
            audioFile={formData.audioFile}
          />
        </div>
        <div className='flex flex-col'>
          <label htmlFor='duration' className='mb-2 font-bold'>
            Select Duration <span className='ml-0.5 text-danger-500'>*</span>
          </label>
          <select
            name='duration'
            id='duration'
            className='rounded bg-surface-100'
            onChange={onChangeDuration}>
            {MEETING_DURATION_OPTIONS.map((option) => (
              <option key={option.value} value={option.value}>
                {option.label}
              </option>
            ))}
          </select>
        </div>
        <div className='flex flex-col'>
          <label htmlFor='guest-select' className='mb-2 font-bold'>
            Select Guest <span className='ml-0.5 text-danger-500'>*</span>
          </label>
          <Select
            aria-label='Select Guest'
            className='custom-react-select-container'
            classNamePrefix='custom-react-select'
            value={formData.guest}
            onChange={onChangeGuest}
            options={options}
            styles={customStyles}
            menuPortalTarget={document.body}
            isSearchable
            menuShouldScrollIntoView={false}
            menuPlacement='auto'
            theme={reactSelectTheme}
            isLoading={isFetchingUsers}
          />
        </div>
        <div className='flex flex-col gap-2'>
          <label htmlFor='date' className='font-bold'>
            Select Date and Time{' '}
            <span className='ml-0.5 text-danger-500'>*</span>
          </label>
          <DatePicker
            id='date'
            className='w-full rounded bg-surface-100'
            selected={formData.date}
            onChange={onChangeDate}
            dateFormat='dd/MM/yyyy hh:mm a'
            showTimeSelect
            minDate={new Date()}
            filterTime={filterPassedTime}
            timeIntervals={15}
          />
        </div>
        <div className='mt-4 flex justify-end gap-4'>
          <OutlinedButton color='danger' onClick={onCancel}>
            Cancel
          </OutlinedButton>
          <Button onClick={handleOpenConfirmModal}>Send Invitation</Button>
        </div>
      </div>
      {userResponse && (
        <AppointmentConfirmModal
          handleClose={handleCloseConfirmModal}
          handleCloseParentModal={onCancel}
          open={isConfirmModalOpen}
          formData={{
            ...formData,
            guestInfo:
              userResponse?.find((u) => u.id === formData.guest.value) ?? null,
          }}
        />
      )}
    </div>
  )
}

export default AppointmentForm
