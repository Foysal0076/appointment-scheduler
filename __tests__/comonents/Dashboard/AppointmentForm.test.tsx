import AppointmentForm from '@/components/Dashboard/AppointmentForm'
import { fireEvent, render, screen } from '@/utils/test-utils'

jest.mock('react-audio-voice-recorder', () => ({
  AudioRecorder: () => <div>Mocked Audio Recorder</div>,
}))

describe('AppointmentForm component', () => {
  const mockProps = {
    userId: '123',
    onCancel: jest.fn(),
  }

  test('renders form fields correctly', () => {
    render(<AppointmentForm {...mockProps} />)
    expect(screen.getByLabelText('Title')).toBeInTheDocument()
    expect(
      screen.getByLabelText((content, element) =>
        content.includes('Description')
      )
    ).toBeInTheDocument()
    expect(screen.getByTestId('audio-message-input')).toBeInTheDocument
    expect(
      screen.getByLabelText((content, element) =>
        content.includes('Select Duration')
      )
    ).toBeInTheDocument()
    expect(screen.getByLabelText('Select Guest')).toBeInTheDocument
    expect(
      screen.getByLabelText((content, element) =>
        content.includes('Select Date and Time')
      )
    ).toBeInTheDocument()
  })

  test('calls onCancel when cancel button is clicked', () => {
    render(<AppointmentForm {...mockProps} />)
    fireEvent.click(screen.getByRole('button', { name: 'Cancel' }))
    expect(mockProps.onCancel).toHaveBeenCalled()
  })

})
