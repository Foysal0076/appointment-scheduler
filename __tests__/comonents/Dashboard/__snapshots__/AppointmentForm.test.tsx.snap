// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppointmentForm component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="w-[85vw] md:max-w-lg"
  >
    <h2
      class="h3 mb-8"
    >
      Create Meeting
    </h2>
    <div
      class="gap- flex flex-col gap-6"
    >
      <div
        class="flex flex-col"
      >
        <label
          class="mb-2 font-bold"
          for="meeting-title"
        >
          Title
        </label>
        <div
          class="custom-input group flex flex-col gap-1"
        >
          <div
            class="relative flex flex-col"
          >
            <input
              class="rounded-sm border-none bg-surface-50 px-4 text-neutral-600 outline-none ring-1 ring-surface-400 placeholder:text-neutral-100 focus:placeholder:text-primary-100 disabled:bg-neutral-30 disabled:text-neutral-100 dark:bg-surface-200 dark:disabled:bg-surface-400  ring-neutral-50 focus:ring-primary-500 dark:focus:ring-primary-300"
              id="meeting-title"
              name="meeting-title"
              placeholder="Give your meeting a title"
            />
          </div>
        </div>
      </div>
      <div
        class="flex flex-col"
      >
        <label
          class="mb-2 font-bold"
          for="meeting-description"
        >
          Description 
          <span
            class="ml-0.5 text-danger-500"
          >
            *
          </span>
        </label>
        <textarea
          class="w-full resize-none rounded bg-surface-50 dark:bg-surface-200"
          id="meeting-description"
          name="meeting-description"
          placeholder="Add a description"
          rows="4"
        />
      </div>
      <div
        class="flex flex-col"
      >
        <label
          class="mb-2 font-bold"
          for="audio-message"
        >
          Audio message
        </label>
        <div
          class="flex flex-col gap-4"
          data-testid="audio-message-input"
          id="audio-message"
        >
          <div>
            Mocked Audio Recorder
          </div>
        </div>
      </div>
      <div
        class="flex flex-col"
      >
        <label
          class="mb-2 font-bold"
          for="duration"
        >
          Select Duration 
          <span
            class="ml-0.5 text-danger-500"
          >
            *
          </span>
        </label>
        <select
          class="rounded bg-surface-100"
          id="duration"
          name="duration"
        >
          <option
            value="30"
          >
            30 minutes
          </option>
          <option
            value="60"
          >
            1 hour
          </option>
          <option
            value="90"
          >
            1 hour 30 minutes
          </option>
          <option
            value="120"
          >
            2 hours
          </option>
          <option
            value="150"
          >
            2 hours 30 minutes
          </option>
          <option
            value="180"
          >
            3 hours
          </option>
        </select>
      </div>
      <div
        class="flex flex-col"
      >
        <label
          class="mb-2 font-bold"
          for="guest-select"
        >
          Select Guest 
          <span
            class="ml-0.5 text-danger-500"
          >
            *
          </span>
        </label>
        <div
          class="custom-react-select-container css-b62m3t-container"
        >
          <span
            class="css-1f43avz-a11yText-A11yText"
            id="react-select-4-live-region"
          />
          <span
            aria-atomic="false"
            aria-live="polite"
            aria-relevant="additions text"
            class="css-1f43avz-a11yText-A11yText"
            role="log"
          />
          <div
            class="custom-react-select__control css-1nlvh3a-control"
          >
            <div
              class="custom-react-select__value-container custom-react-select__value-container--has-value css-hlgwow"
            >
              <div
                class="custom-react-select__single-value css-1dimb5e-singleValue"
              />
              <div
                class="custom-react-select__input-container css-19bb58m"
                data-value=""
              >
                <input
                  aria-activedescendant=""
                  aria-autocomplete="list"
                  aria-expanded="false"
                  aria-haspopup="true"
                  aria-label="Select Guest"
                  autocapitalize="none"
                  autocomplete="off"
                  autocorrect="off"
                  class="custom-react-select__input"
                  id="react-select-4-input"
                  role="combobox"
                  spellcheck="false"
                  style="opacity: 1; width: 100%; grid-area: 1 / 2; min-width: 2px; border: 0px; margin: 0px; outline: 0; padding: 0px;"
                  tabindex="0"
                  type="text"
                  value=""
                />
              </div>
            </div>
            <div
              class="custom-react-select__indicators css-1wy0on6"
            >
              <div
                aria-hidden="true"
                class="custom-react-select__indicator custom-react-select__loading-indicator css-pogzpp-loadingIndicator"
              >
                <span
                  class="css-1h1juvu-LoadingDot"
                />
                <span
                  class="css-1swgd6a-LoadingDot"
                />
                <span
                  class="css-41hts2-LoadingDot"
                />
              </div>
              <span
                class="custom-react-select__indicator-separator css-1u9des2-indicatorSeparator"
              />
              <div
                aria-hidden="true"
                class="custom-react-select__indicator custom-react-select__dropdown-indicator css-1xc3v61-indicatorContainer"
              >
                <svg
                  aria-hidden="true"
                  class="css-tj5bde-Svg"
                  focusable="false"
                  height="20"
                  viewBox="0 0 20 20"
                  width="20"
                >
                  <path
                    d="M4.516 7.548c0.436-0.446 1.043-0.481 1.576 0l3.908 3.747 3.908-3.747c0.533-0.481 1.141-0.446 1.574 0 0.436 0.445 0.408 1.197 0 1.615-0.406 0.418-4.695 4.502-4.695 4.502-0.217 0.223-0.502 0.335-0.787 0.335s-0.57-0.112-0.789-0.335c0 0-4.287-4.084-4.695-4.502s-0.436-1.17 0-1.615z"
                  />
                </svg>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="flex flex-col gap-2"
      >
        <label
          class="font-bold"
          for="date"
        >
          Select Date and Time 
          <span
            class="ml-0.5 text-danger-500"
          >
            *
          </span>
        </label>
        <div
          class="react-datepicker-wrapper"
        >
          <div
            class="react-datepicker__input-container"
          >
            <input
              class="w-full rounded bg-surface-100"
              id="date"
              type="text"
              value="08/08/2024 06:30 AM"
            />
          </div>
        </div>
      </div>
      <div
        class="mt-4 flex justify-end gap-4"
      >
        <button
          class="relative select-none rounded bg-none border-2 focus:ring-4 disabled:active:transform-none disabled:cursor-not-allowed dark:hover:bg-neutral-20/50 transition-all duration-300 text-nowrap focus:outline-none font-medium active:scale-[.9] hover:bg-danger-50 focus:ring-danger-100 border-danger-500/80 text-danger-500  focus:ring-danger-100 dark:text-danger-200 dark:border-danger-200 disabled:text-danger-200 disabled:border-danger-100 disabled:hover:bg-transparent dark:disabled:bg-danger-50/20 dark:disabled:border-danger-50/10 dark:disabled:text-danger-100/60 h-[2.5rem] text-md px-4"
        >
          <div
            class="visible font-Rubik"
          >
            Cancel
          </div>
        </button>
        <button
          class="relative select-none rounded text-white focus:ring-4 disabled:active:transform-none active:scale-[.9] disabled:cursor-not-allowed transition-all duration-300 text-nowrap focus:outline-none font-medium bg-primary-500 disabled:text-primary-200 disabled:bg-primary-50 hover:bg-primary-600 focus:ring-primary-100 focus:outline-primary-200  h-[2.5rem] text-md px-4"
        >
          <div
            class="visible font-Rubik"
          >
            Send Invitation
          </div>
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
